module_utils/network_lsr/argument_validator.py

```diff
@@ -1802,3 +1802,23 @@ def validate_connection_one(self, mode, connections, idx):
                     idx,
                     "ip.ipv6_disabled is not supported by initscripts.",
                 )
+        # when IPv4 or IPv6 is disabled, settings for ip.dns_options and
+        # ip.dns_search are not allowed
+        if connection["ip"]["dns_options"] or connection["ip"]["dns_search"]:
+            addrs4 = list(
+                [
+                    address
+                    for address in connection["ip"]["address"]
+                    if address["family"] == socket.AF_INET
+                ]
+            )
+            if not connection["ip"]["dhcp4"] and not addrs4:
+                raise ValidationError.from_connection(
+                    idx,
+                    "IPv4 needs to be enabled to support IPv4 nameservers.",
+                )
+            if connection["ip"]["ipv6_disabled"]:
+                raise ValidationError.from_connection(
+                    idx,
+                    "IPv6 needs to be enabled to support IPv6 nameservers.",
```

Comment:
the error message talks about "nameservers", but here we are inside a
```
if connection["ip"]["dns_options"] or connection["ip"]["dns_search"]:
```
block. That is, it does not check the nameservers, but other DNS settings.

More importantly, setting "dns_search" and/or "dns_option" is fine if **any** address family is enabled. The current code would reject setting these DNS settings, if any of the address families is disabled.

Reply:
I do not understand "More importantly, setting "dns_search" and/or "dns_option" is fine if any address family is **enabled**. " Should this be **disabled** instead?

Reply:
it is OK set set dns-search/dns-options, if at least one of the address families is enabled.

The current code checks that both are enabled.

Reply:
I just realized the validation for dns and dns_search/dns_options should be different:
```
        # Setting ip.dns is not allowed when corresponding IP method for that
        # nameserver is disabled
        for nameserver in connection["ip"]["dns"]:
            has_addrs4 = any(
                address["family"] == socket.AF_INET
                for address in connection["ip"]["address"]
            )
            if nameserver["family"] == socket.AF_INET and (
                not connection["ip"]["dhcp4"] and not has_addrs4
            ):
                raise ValidationError.from_connection(
                    idx,
                    "IPv4 needs to be enabled to support IPv4 nameservers.",
                )
            if (
                nameserver["family"] == socket.AF_INET6
                and connection["ip"]["ipv6_disabled"]
            ):
                raise ValidationError.from_connection(
                    idx,
                    "IPv6 needs to be enabled to support IPv6 nameservers.",
                )
        # when IPv4 and IPv6 are disabled, setting ip.dns_options or
        # ip.dns_search is not allowed
        if connection["ip"]["dns_search"] or connection["ip"]["dns_options"]:
            has_addrs4 = any(
                address["family"] == socket.AF_INET
                for address in connection["ip"]["address"]
            )
            if (not connection["ip"]["dhcp4"] and not has_addrs4) and connection["ip"][
                "ipv6_disabled"
            ]:
                raise ValidationError.from_connection(
                    idx,
                    "Setting 'dns_search' or 'dns_options' is not allowed when "
                    "both IPv4 and IPv6 are disabled.",
                )
```